### PR TITLE
Added a couple more symint magic methods + symbolic shape infra

### PR DESCRIPTION
**Pull request by Chillee**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #81093
* #81092
* #81088
* __->__ #81086
* #81078

### CODE REVIEW

File: torch/fx/experimental/symbolic_shapes.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
+import torch
+
+try:
+    import sympy
+    HAS_SYMPY = True
+except ImportError:
+    HAS_SYMPY = False
```
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:7]: This is just code motion, I hope?

**Review by ezyang — APPROVED** [review]

File: torch/fx/experimental/symbolic_shapes.py (hunk 1 above)
Comment (Chillee) [reply torch/fx/experimental/symbolic_shapes.py:7]: wdym? This is here so that if a file imports this without having sympy installed, it doesn't error.

Comment (ezyang) [reply torch/fx/experimental/symbolic_shapes.py:7]: oh I wanted to know if the code here had changed

Comment (Chillee) [reply torch/fx/experimental/symbolic_shapes.py:7]: Oh, I see. Yes, it's changed a little bit. Nick previously copied the POC from a previous version into the test file, I added an updated version into this file that we'll be building off moving forward, and modified Nick's tests to use it.

File: torch/fx/experimental/symbolic_shapes.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+class PySymInt(object):
```
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:9]: would be nice to have docblocks

File: torch/fx/experimental/symbolic_shapes.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    def __str__(self):
+        return f"PySymInt({self.expr})"
+
+    def __int__(self):
+        raise RuntimeError("Trying to extract a concrete int out of a symbolic int")
+
+        return self.shape_env.evaluate_expr(self.expr)
+
+    def __bool__(self):
+        return bool(self.shape_env.evaluate_expr(self.expr))
```
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:26]: looks inconsistent 🧐

File: torch/fx/experimental/symbolic_shapes.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+
+reflectable_magic_methods = {
```
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:29]: not obvious what reflectable means here

File: torch/fx/experimental/symbolic_shapes.py [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+            return PySymInt(func(self.expr, other), self.shape_env)
+        return magic_impl
+
+    # this should be wrapped transparently into torch._C.SymbolicIntNode
+    setattr(PySymInt, method_name, create_magic_impl(func))
+
+class ShapeEnv(object):
+    def __init__(self):
+        self.guards = []
+        self.shape_env = {}
```
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:61]: This being a dict means that entries in the shape env never die, even if the sympy expression becomes unreachable

File: torch/fx/experimental/symbolic_shapes.py [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+            return val
+        sympy_expr = sympy.Symbol(name, positive=True)
+        py_sym_int = PySymInt(sympy_expr, self)
+        cpp_sym_int = torch._C.SymbolicIntNode.new_symint(py_sym_int)  # type: ignore[attr-defined]
+        self.shape_env[sympy_expr] = val
+        return cpp_sym_int
+
+    def try_constantify(self, expr):
+        # Simplifies assuming that shape vars > 0
+        new_shape_env = {k: sympy.Symbol(f'shape_{idx}', positive=True) + 1 for idx, k in enumerate(self.shape_env.keys())}
```
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:76]: does sympy positive mean >= 0? weird

File: torch/fx/experimental/symbolic_shapes.py [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+        new_expr = expr.subs(new_shape_env)
+        new_expr = new_expr.simplify()
+        if len(list(new_expr.free_symbols)) == 0:
+            return new_expr
+        return None
+
+    def evaluate_expr(self, expr):
+        if (const_expr := self.try_constantify(expr)) is not None:
```
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:84]: our python isn't new enough for walrus is it?

File: torch/fx/experimental/symbolic_shapes.py (hunk 6 above)
Comment (Chillee) [reply torch/fx/experimental/symbolic_shapes.py:76]: Oops, comment is wrong.

File: torch/fx/experimental/symbolic_shapes.py (hunk 7 above)
Comment (Chillee) [reply torch/fx/experimental/symbolic_shapes.py:84]: yeah, it isn't, which is why tests are failing :'(

File: torch/fx/experimental/symbolic_shapes.py (hunk 4 above)
Comment (Chillee) [reply torch/fx/experimental/symbolic_shapes.py:29]: Just that `__radd__` exists for `__add__`, will add a comment.

File: torch/fx/experimental/symbolic_shapes.py (hunk 3 above)
Comment (Chillee) [reply torch/fx/experimental/symbolic_shapes.py:26]: Yeah I'll add a comment. But basically, coercing to bool is generally reasonable. Coercing to int, on the other hand, is a very easy way of shooting yourself in the foot.

Will add a comment.

Later we'll probably need a way to say "I know what I'm doing, please coerce to an int", but that's for later :)

File: torch/fx/experimental/symbolic_shapes.py (hunk 5 above)
Comment (Chillee) [reply torch/fx/experimental/symbolic_shapes.py:61]: is that an issue?

Comment (ezyang) [reply torch/fx/experimental/symbolic_shapes.py:66]: it is always good to practice good memory hygiene :)

**Pull request closed without merging**